rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+
+        // Update heap memory with new association. Little endian.
+        // (write u16 to slot i in new_pages.)
+        self.0.set(i * 2, block_id_lower);
+        self.0.set(i * 2 + 1, block_id_upper);
+    }
+
+    pub unsafe fn get_ith_block_id(&self, i: u32) -> BlockId {
+        let lower: u16 = self.0.get(i * 2) as u16;
+        let upper: u16 = self.0.get(i * 2 + 1) as u16;
+        BlockId(upper << 8 | lower)
+    }
+}
+
+impl RegionObject {
+    pub unsafe fn from_value(v: &Value) -> Self {
+        RegionObject(v.as_region())
+    }
+
+    pub unsafe fn id(&self) -> RegionId {
+        RegionId((*self.0).id)
+    }
+
+    pub unsafe fn check_relative_into_absolute_offset(&self, offset: u64) {
+        if !((offset / meta_data::size::PAGE_IN_BYTES) < (*self.0).page_count.into()) {
+            rts_trap_with("region access out of bounds.");
+        };
+    }
+
+    // computes absolute offset, BlockId, and remaining length (of the
+    // given block) for a relative offset.
+    //
+    // assumes but does not run check_relative_into_absolute_offset for
+    // this offset, or some higher one.
+    pub unsafe fn relative_into_absolute_info(&self, offset: u64) -> (u64, BlockId, u64) {
+        let av = AccessVector::from_value(&(*self.0).vec_pages);
+
+        // Which block (rank relative to this region)?
+        let block_rank = offset / meta_data::size::BLOCK_IN_BYTES;
+
+        // Where in that block?
+        let intra_block_index = offset % meta_data::size::BLOCK_IN_BYTES;
+
+        // Where is that block located in stable memory (global rank)?
+        let block_id = av.get_ith_block_id(block_rank as u32);
+
+        // address of the byte to load from stable memory:
+        let offset = meta_data::offset::BLOCK_ZERO
+            + block_id.0 as u64 * meta_data::size::BLOCK_IN_BYTES
+            + intra_block_index;
+        (
+            offset,
+            block_id,
+            meta_data::size::BLOCK_IN_BYTES - intra_block_index,
+        )
+    }
+
+    // compute the absolute offset, begin block, remaining length of
+    // begin block, and end block for a relative offset and length.
+    //
+    // NB: BlockIds can be used to do case analysis (same or diff
+    // block?) when planning successive reads/writes.
+    pub unsafe fn relative_into_absolute_span(
+        &self,
+        offset: u64,
+        len: u64,
+    ) -> (u64, BlockId, u64, BlockId) {
+        let (off, b1, b1_len) = self.relative_into_absolute_info(offset);
+        if len == 0 {
+            return (off, b1.clone(), b1_len, b1);
+        };
+        let final_offset = offset + len - 1;
+        let (_, b2, _) = self.relative_into_absolute_info(final_offset);
+        (off, b1, b1_len, b2)
+    }
+}
+
+// Mutable meta data stored in stable memory header (See motoko/design/StableRegions.md)
+mod meta_data {
+
+    /// Maximum number of entities.
+    pub mod max {
+        pub const BLOCKS: u16 = 32768;
+        pub const REGIONS: u16 = 32767;
```

Comment:
Detail (just an idea, only cosmetics):
```suggestion
        pub const BLOCKS: u16 = 32 * 1024;
        pub const REGIONS: u16 = 32 * 1024 - 1;
```